phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/UpdateExpressionUtils.java

```diff
@@ -0,0 +1,794 @@
+      }
+      curIdx++;
+      int arrayIdx = Integer.parseInt(arrayIdxStr.toString());
+      if (value == null || !value.isArray()) {
+        LOGGER.error("Value is null or not document. Value: {}, Idx: {}, fieldKey: {}, New val: {}",
+            value, idx, fieldKey, newVal);
+        throw new RuntimeException("Value is null or not array.");
+      }
+      BsonArray nestedList = (BsonArray) value;
+      if (curIdx == fieldKey.length()) {
+        // reached leaf node of the document tree, update the value as per the update operator
+        // semantics
+        updateValueAtLeafNode(fieldKey, newVal, updateOp, arrayIdx, nestedList);
+        return;
+      }
+      BsonValue nestedValue = nestedList.get(arrayIdx);
+      if (nestedValue == null) {
+        LOGGER.error("Should have found nested list for index {}", arrayIdx);
+        return;
+      }
+      updateNestedField(nestedValue, curIdx, fieldKey, newVal, updateOp);
+    } else {
+      StringBuilder sb = new StringBuilder();
+      for (int i = idx; i < fieldKey.length(); i++) {
+        if (fieldKey.charAt(i) == '.') {
+          BsonValue topFieldValue = ((BsonDocument) value).get(sb.toString());
+          if (topFieldValue == null) {
+            LOGGER.error("Incorrect access. Should have found nested bsonDocument for {}", sb);
+            throw new RuntimeException("Document does not contain key: " + sb);
+          }
+          updateNestedField(topFieldValue, i, fieldKey, newVal, updateOp);
+          return;
+        } else if (fieldKey.charAt(i) == '[') {
+          BsonValue topFieldValue = ((BsonDocument) value).get(sb.toString());
+          if (topFieldValue == null) {
+            LOGGER.error("Incorrect access. Should have found nested list for {}", sb);
+            throw new RuntimeException("Document does not contain key: " + sb);
+          }
+          updateNestedField(topFieldValue, i, fieldKey, newVal, updateOp);
+          return;
+        } else {
+          sb.append(fieldKey.charAt(i));
+        }
+      }
+    }
+  }
+
+  /**
+   * Perform update operation at the leaf node. This method is called only when the leaf
+   * node or the target node for the given field key is encountered while iterating through
+   * the tree structure. This is specifically called when the closest ancestor of the
+   * given node is a nested array.
+   *
+   * @param fieldKey The full field key.
+   * @param newVal New value to be used while performing update operation on the existing node.
+   * @param updateOp Type of the update operation to be performed.
+   * @param arrayIdx The index of the array at which the target node value is to be updated.
+   * @param nestedList The parent or ancestor node, expected to be nested array only.
+   * For example, for "category.subcategories.brands[5]" field key, the nestedList is expected to
+   * be "brands" array and arrayIdx is expected to be 5.
+   */
+  private static void updateValueAtLeafNode(final String fieldKey,
```

Comment:
This should be called updateArray instead of updateValue as this is specific to an Array field.